pkg/apis/pipeline/v1beta1/when_types.go

```diff
@@ -0,0 +1,142 @@
+	for i := range we.Values {
+		if we.Values[i] == we.Input {
+			return true
+		}
+	}
+	return false
+}
+
+func (we *WhenExpression) isTrue() bool {
+	switch we.Operator {
+	case selection.In:
+		return we.isInputInValues()
+	case selection.NotIn:
+		return !we.isInputInValues()
+	default:
+		// should not reach here, we validate that operators
+		// are selection.In and selection.NotIn only
+		return false
+	}
+}
+
+func (we *WhenExpression) hasVariable() bool {
+	if _, hasVariable := we.GetVarSubstitutionExpressions(); hasVariable {
+		return true
+	}
+	return false
+}
+
+func (we *WhenExpression) applyReplacements(replacements map[string]string) WhenExpression {
+	replacedInput := ApplyReplacements(we.Input, replacements)
+
+	var replacedValues []string
+	for _, val := range we.Values {
+		replacedValues = append(replacedValues, ApplyReplacements(val, replacements))
+	}
+
+	return WhenExpression{Input: replacedInput, Operator: we.Operator, Values: replacedValues}
+}
+
+// WhenExpressions are used to specify whether a Task should be executed or skipped.
+type WhenExpressions []WhenExpression
+
+// Allows evaluates an Input's relationship to an array of Values, based on the Operator
+// to determine whether all the When Expressions are True.
+func (wes WhenExpressions) Allows() bool {
+	for _, we := range wes {
+		if !we.isTrue() {
+			return false
+		}
+	}
+	return true
+}
+
+// HaveVariables indicates whether When Expressions contains variables, such as Parameters
+// or Results in the Inputs or Values.
+func (wes WhenExpressions) HaveVariables() bool {
+	for _, we := range wes {
+		if we.hasVariable() {
+			return true
+		}
+	}
+	return false
+}
+
+// ReplaceWhenExpressionsVariables interpolates variables, such as Parameters and Results, in
+// the Input and Values.
+func (wes WhenExpressions) ReplaceWhenExpressionsVariables(replacements map[string]string) WhenExpressions {
+	var replaced []WhenExpression
+	for _, we := range wes {
+		replaced = append(replaced, we.applyReplacements(replacements))
+	}
+	return replaced
+}
+
+// WhenExpressionsResult contains the When Expression that was evaluated and
+// a boolean indicating whether the When Expression is true or false based on its
+// Input's relationship to its values as specified by its Operator.
+type WhenExpressionsResult struct {
+	// Expression is the expression, made up of Input, Operator and Values.
+	Expression *WhenExpression `json:"expression"`
+	// IsTrue is a boolean from evaluating the Expression.
+	IsTrue bool `json:"isTrue"`
+}
+
+// GetStatus iterates through the When Expressions, and evaluates an Input's
+// relationship to an array of Values based on the Operator and produces a Status update
+// for Tasks with When Expressions.
+func (wes WhenExpressions) GetStatus() []*WhenExpressionsResult {
+	var status []*WhenExpressionsResult
+	for i := range wes {
+		status = append(status, &WhenExpressionsResult{
+			Expression: &wes[i],
+			IsTrue:     wes[i].isTrue(),
```

Comment:
`IsTrue` is not necessary here, at this point (with resolved expressions), its very clear that `foo` is in `["foo", "bar"]` and `zoo` is not in `["foo", "bar"]`, its consuming unnecessary space which might be fine for a pipeline with few tasks but might not be fine running it repeatedly or for a pipeline with thousands of tasks 😜 

```
    "skip-this-task": [
                        {
                            "expression": {
                                "Input": "zoo",
                                "Operator": "in",
                                "Values": [
                                    "foo"
                                ]
                            },
                            "isTrue": false
                        }
                    ]
```

Reply:
I was initially thinking that the values could be a long list and someone might find it tedious to look through the list, but I think that's a good point...

But that also makes me question whether we need the expressions because inputs & values in the pipelinespec could be sufficient because:
- the non-variables are right there in the spec
- the param variables in the when expressions in the pipeline spec can be looked up in the pipeline spec
- the result variables in the when expressions in the pipeline spec can be looked up in result section of the relevant taskrun

so maybe I can get remove the When Expressions Status from TaskRuns and Skipped, since all the details they need are already somewhere in the status 🤔

Reply:
im a bit torn! if we want, we could leave this out entirely for now and add it later, but im slightly inclined toward including it because folks will want to understand why something got skipped - i guess they can reverse engineer it from the information that is already there, but one difference between when expressions and tasks is that the logic in a task is up to the task itself, while the logic for a when expression is in the controller, so i can see being a bit more verbose about "this is exactly what we computed and this was the result"

Reply:
@bobcatfish yup I can understand your concern, how about storing these details in controller logs? Do we expect users would have access to the logs in standard CI/CD setup?